### PR TITLE
Integrations Boilerplate Codegen Tool

### PR DESCRIPTION
**Pull request by sentilesdal**
Adds a codegen tool to generate much of boilerplate code necessary for new integrations of hyperdrive.  The overall goal is to be to just define a simple `config.yaml` file and then be able to run:

```bash
hyperdrive-codegen --config path/to/config.yaml --out contracts/src
```

This will generate a slew of files:

```
deployers/
    testeth/
      TestETHHyperdriveCoreDeployer.sol
      TestETHHyperdriveDeployerCoordinator.sol
      TestETHTarget0Deployer.sol
      TestETHTarget1Deployer.sol
      TestETHTarget2Deployer.sol
      TestETHTarget3Deployer.sol
      TestETHTarget4Deployer.sol

instances/
    testeth/
      TestETHBase.sol
      TestETHHyperdrive.sol
      TestETHTarget0.sol
      TestETHTarget1.sol
      TestETHTarget2.sol
      TestETHTarget3.sol
      TestETHTarget4.sol

interfaces/
      ITestETH.sol
      ITestETHHyperdrive.sol
```

The configuration yaml will allow us to create certain preconfigured defaults, such as disabling base deposits and stubbing out default functions for those.

I've also added a github workflow to test the different configurations by making sure all the generated code compiles.

### CODE REVIEW

File: codegen/templates/interfaces/IHyperdrive.sol.jinja
Comment (cashd) [inline codegen/templates/interfaces/IHyperdrive.sol.jinja:6]: I don't think this prettier ignore is needed

**Review by cashd — APPROVED** [review]
Great job this is going to be soo useful! can't wait to use it in the next integration. I ran through the readme and got every working very smoothly.

File: codegen/templates/interfaces/IYieldSource.sol.jinja
Comment (jalextowle) [inline codegen/templates/interfaces/IYieldSource.sol.jinja:1]: We should always use Apache-2.0. We can write interfaces by hand if we need to. It's important that we don't bring any copy-left code into the Hyperdrive codebase.

**Review by jalextowle — COMMENTED** [review]
This looks great, really nice job dude! Aside from the licensing comment, I have one other high-level comment. I think it would be good to keep the `interfaces/` directory flattened for consistency. If we're going to add folders, we should add folders everywhere. Since that isn't something that we've done for the old yield sources, I'd prefer updating this to output the autogenerated interface into the `contracts/src/interfaces`.

File: codegen/templates/interfaces/IYieldSource.sol.jinja
Comment (sentilesdal) [reply codegen/templates/interfaces/IYieldSource.sol.jinja:1]: roger, this was copied from our IER4626.sol, should we update there as well?

**Comment by sentilesdal** [comment]
@jalextowle flattened interfaces/ files and updated the license.  left a comment about the GPL license in IERC20.sol

**Review by jalextowle — APPROVED** [review]
Looks great! I think the last thing you should do is add yourself and cash as codeowners for the `codegen/` folder so that you can make changes to the tool without needing a review from me or Jonny. We'll still need to review new instances for now.

**Pull request merged**
